bundles/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/action/ZmActions.java

```diff
@@ -0,0 +1,136 @@
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.zoneminder.action;
+
+import java.lang.reflect.Method;
+import java.lang.reflect.Proxy;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.openhab.binding.zoneminder.internal.handler.ZmMonitorHandler;
+import org.openhab.core.automation.annotation.ActionInput;
+import org.openhab.core.automation.annotation.RuleAction;
+import org.openhab.core.thing.binding.ThingActions;
+import org.openhab.core.thing.binding.ThingActionsScope;
+import org.openhab.core.thing.binding.ThingHandler;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link ZmActions} defines the thing actions provided by this binding.
+ *
+ * <b>Note:</b>The static method <b>invokeMethodOf</b> handles the case where
+ * the test <i>actions instanceof ZmActions</i> fails. This test can fail
+ * due to an issue in openHAB core v2.5.0 where the {@link ZmActions} class
+ * can be loaded by a different classloader than the <i>actions</i> instance.
+ *
+ * @author Mark Hilbush - Initial contribution
+ */
+@ThingActionsScope(name = "zm")
+@NonNullByDefault
+public class ZmActions implements ThingActions, IZmActions {
+    private final Logger logger = LoggerFactory.getLogger(ZmActions.class);
+
+    private @Nullable ZmMonitorHandler handler;
+
+    @Override
+    public @Nullable ThingHandler getThingHandler() {
+        return this.handler;
+    }
+
+    @Override
+    public void setThingHandler(@Nullable ThingHandler handler) {
+        if (handler instanceof ZmMonitorHandler) {
+            this.handler = (ZmMonitorHandler) handler;
+        }
+    }
+
+    private static IZmActions invokeMethodOf(@Nullable ThingActions actions) {
+        if (actions == null) {
+            throw new IllegalArgumentException("actions cannot be null");
+        }
+        if (actions.getClass().getName().equals(ZmActions.class.getName())) {
+            if (actions instanceof IZmActions) {
+                return (IZmActions) actions;
+            } else {
+                return (IZmActions) Proxy.newProxyInstance(IZmActions.class.getClassLoader(),
+                        new Class[] { IZmActions.class }, (Object proxy, Method method, Object[] args) -> {
+                            Method m = actions.getClass().getDeclaredMethod(method.getName(),
+                                    method.getParameterTypes());
+                            return m.invoke(actions, args);
+                        });
+            }
+        }
+        throw new IllegalArgumentException("Actions is not an instance of ZmActions");
+    }
```

Comment:
Since DSL rules will still be supported in OH3, I don't recall seeing anything that fixes the root cause of the issue. And the [issue in core](https://github.com/openhab/openhab-core/issues/1265) is still open. But @cweitkamp would know for sure. Let wait for his reply.

Reply:
Please correct me if I am wrong, I am not really up-to-date right now: IIRC there was no real confirmation for `ThingActions` are working correctly in OH3 NGRE and we wanted to keep those workarounds and bulk remove them later.